File: agenkit/patterns/task.py
"""
Task pattern for one-shot agent execution with lifecycle management.

This module provides the Task pattern, which wraps an Agent for single-use
execution with automatic resource cleanup.
"""

import asyncio
from typing import Any

from agenkit.interfaces import Agent, Message


class Task:
    """
    One-shot agent execution with lifecycle management.

    A Task wraps an Agent for single-use execution, providing:
    - Explicit one-shot semantics
    - Automatic resource cleanup
    - Task-specific middleware (timeout, retries)
    - Lifecycle hooks (before_task, after_task)
    - Prevention of reuse after completion

    The key distinction from Agent:
    - **Agent**: Multi-turn conversation with state
    - **Task**: One-shot execution, then cleanup

    Args:
        agent: The Agent to execute
        timeout: Optional timeout in seconds for task execution
        retries: Number of retry attempts on failure (default: 0)
        **kwargs: Additional configuration

    Example:
        >>> from agenkit.patterns import Task
        >>> from agenkit import Agent, Message
        >>>
        >>> # Basic usage
        >>> async def summarize_document(agent: Agent, document: str) -> str:
        ...     task = Task(agent)
        ...     try:
        ...         messages = [Message(role="user", content=f"Summarize: {document}")]
        ...         result = await task.execute(messages)
        ...         return result.content
        ...     finally:
        ...         await task.cleanup()

    Context manager usage (recommended):
        >>> async with Task(agent) as task:
        ...     result = await task.execute(messages)
        ...     # Automatic cleanup on exit

    With timeout and retries:
        >>> async with Task(agent, timeout=30.0, retries=2) as task:
        ...     result = await task.execute(messages)

    Pattern comparison:
        - Use Task when: Single purpose operation that needs cleanup
        - Use Agent when: Multi-turn conversation with state
        - Examples: summarize_document, classify_text, extract_entities
    """

    def __init__(
        self,
        agent: Agent,
        timeout: float | None = None,
        retries: int = 0,
        **kwargs: Any,
    ):
        """
        Initialize a Task.

        Args:
            agent: The Agent to execute
            timeout: Optional timeout in seconds
            retries: Number of retry attempts on failure
            **kwargs: Additional configuration
        """
        self._agent = agent
        self._timeout = timeout
        self._retries = retries
        self._config = kwargs
        self._completed = False
        self._result: Message | None = None

    async def execute(
        self,
        messages: list[Message],
        **kwargs: Any,
    ) -> Message:
        """
        Execute the task once.

        This method can only be called once per Task instance. After execution
        completes (successfully or with error), the Task is marked as completed
        and cannot be reused.

        Args:
            messages: Input messages for the agent
            **kwargs: Additional arguments passed to agent.call()

        Returns:
            The agent's response as a Message

        Raises:
            RuntimeError: If the task has already been completed
            asyncio.TimeoutError: If execution exceeds the timeout
            Exception: Any exception raised by the agent

        Example:
            >>> task = Task(agent, timeout=30.0, retries=2)
            >>> result = await task.execute(messages)
            >>> # Cannot call execute() again - will raise RuntimeError
        """
        if self._completed:
            raise RuntimeError(
                "Task already completed. Create a new Task for another execution."
            )

        # Merge kwargs with config
        call_kwargs = {**self._config, **kwargs}

        # Execute with retries
        attempts = self._retries + 1  # retries=0 means 1 attempt
        last_error = None

        for attempt in range(attempts):
            try:
                # Execute with optional timeout
                if self._timeout:
                    result = await asyncio.wait_for(
                        self._agent.call(messages, **call_kwargs),
                        timeout=self._timeout,
                    )
                else:
                    result = await self._agent.call(messages, **call_kwargs)

                # Success - mark completed and return
                self._completed = True
                self._result = result
                return result

            except asyncio.TimeoutError:
                # Timeout - don't retry
                self._completed = True
                await self.cleanup()
                raise

            except Exception as e:
                last_error = e
                # If this was the last attempt, fail
                if attempt == attempts - 1:
                    self._completed = True
                    await self.cleanup()
                    raise
                # Otherwise, retry after a brief delay
                await asyncio.sleep(0.1 * (attempt + 1))  # Exponential backoff

        # Should never reach here, but just in case
        self._completed = True
        await self.cleanup()
        raise last_error or RuntimeError("Task execution failed")

    async def cleanup(self) -> None:
        """
        Clean up resources after task completion.

        This method is called automatically when:
        - Task execution completes successfully
        - Task execution fails with an error
        - Using the Task as a context manager (on __aexit__)

        Override this method in subclasses to add custom cleanup logic:
        - Close network connections
        - Release memory/resources
        - Save state to disk
        - Send telemetry

        Example:
            >>> class CustomTask(Task):
            ...     async def cleanup(self) -> None:
            ...         # Custom cleanup logic
            ...         await self._agent.close()
            ...         await super().cleanup()
        """
        # Default implementation - hook for subclasses
        # In the future, this could:
        # - Close agent connections
        # - Release middleware resources
        # - Log completion metrics
        pass

    @property
    def completed(self) -> bool:
        """
        Check if the task has been completed.

        Returns:
            True if execute() has been called, False otherwise
        """
        return self._completed

    @property
    def result(self) -> Message | None:
        """
        Get the result of the task execution.

        Returns:
            The result Message if execution completed successfully, None otherwise
        """
        return self._result

    async def __aenter__(self) -> "Task":
        """
        Enter the context manager.

        Returns:
            The Task instance for use in the with block
        """
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: Any,
    ) -> None:
        """
        Exit the context manager and perform cleanup.

        Args:
            exc_type: Exception type if an exception occurred
            exc_val: Exception value if an exception occurred
            exc_tb: Exception traceback if an exception occurred
        """
        await self.cleanup()

File: agenkit-ts/examples/src/patterns/task.js
"use strict";
/**
 * Task pattern for one-shot agent execution with lifecycle management.
 *
 * This module provides the Task pattern, which wraps an Agent for single-use
 * execution with automatic resource cleanup.
 *
 * Key Features:
 * - One-shot execution semantics
 * - Automatic resource cleanup
 * - Timeout support
 * - Retry logic with exponential backoff
 * - Prevention of reuse after completion
 * - Context manager support (async)
 *
 * Example:
 * ```typescript
 * // Basic usage
 * const task = new Task(agent, { timeout: 30000, retries: 2 });
 * try {
 *   const result = await task.execute(message);
 *   console.log(result.content);
 * } finally {
 *   await task.cleanup();
 * }
 *
 * // With context manager pattern
 * await Task.withTask(agent, async (task) => {
 *   const result = await task.execute(message);
 *   return result;
 * });
 * ```
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeoutError = exports.Task = void 0;
exports.executeTask = executeTask;
/**
 * One-shot agent execution with lifecycle management.
 *
 * A Task wraps an Agent for single-use execution, providing:
 * - Explicit one-shot semantics
 * - Automatic resource cleanup
 * - Task-specific configuration (timeout, retries)
 * - Prevention of reuse after completion
 *
 * Key Distinction:
 * - **Agent**: Multi-turn conversation with state
 * - **Task**: One-shot execution, then cleanup
 *
 * Use Task when:
 * - Single purpose operation that needs cleanup
 * - You want explicit resource management
 * - You need timeout/retry at task level
 *
 * Examples: summarize_document, classify_text, extract_entities
 */
class Task {
    constructor(agent, config = {}) {
        this.agent = agent;
        this.timeout = config.timeout;
        this.retries = config.retries || 0;
        this.config = config;
        this._completed = false;
    }
    /**
     * Execute the task once.
     *
     * This method can only be called once per Task instance. After execution
     * completes (successfully or with error), the Task is marked as completed
     * and cannot be reused.
     *
     * @param message Input message for the agent
     * @returns The agent's response
     * @throws RuntimeError if task already completed
     * @throws TimeoutError if execution exceeds timeout
     */
    async execute(message) {
        if (this._completed) {
            throw new Error('Task already completed. Create a new Task for another execution.');
        }
        const attempts = this.retries + 1; // retries=0 means 1 attempt
        let lastError;
        for (let attempt = 0; attempt < attempts; attempt++) {
            try {
                // Execute with optional timeout
                const result = this.timeout
                    ? await this.withTimeout(this.agent.process(message), this.timeout)
                    : await this.agent.process(message);
                // Success - mark completed and return
                this._completed = true;
                this._result = result;
                return result;
            }
            catch (error) {
                lastError = error instanceof Error ? error : new Error(String(error));
                // If timeout, don't retry
                if (error instanceof TimeoutError) {
                    this._completed = true;
                    await this.cleanup();
                    throw error;
                }
                // If this was the last attempt, fail
                if (attempt === attempts - 1) {
                    this._completed = true;
                    await this.cleanup();
                    throw error;
                }
                // Otherwise, retry after exponential backoff
                await this.sleep(100 * (attempt + 1));
            }
        }
        // Should never reach here, but just in case
        this._completed = true;
        await this.cleanup();
        throw lastError || new Error('Task execution failed');
    }
    /**
     * Clean up resources after task completion.
     *
     * This method is called automatically when:
     * - Task execution completes successfully
     * - Task execution fails with an error
     * - Using withTask() (automatic cleanup)
     *
     * Override this method in subclasses to add custom cleanup logic:
     * - Close network connections
     * - Release memory/resources
     * - Save state to disk
     * - Send telemetry
     */
    async cleanup() {
        // Default implementation - hook for subclasses
        // Could close agent connections, release middleware resources, etc.
    }
    /**
     * Check if the task has been completed.
     */
    get completed() {
        return this._completed;
    }
    /**
     * Get the result of the task (if completed successfully).
     */
    get result() {
        return this._result;
    }
    /**
     * Execute a function with timeout.
     */
    async withTimeout(promise, timeoutMs) {
        return Promise.race([
            promise,
            new Promise((_, reject) => setTimeout(() => reject(new TimeoutError(`Task timed out after ${timeoutMs}ms`)), timeoutMs)),
        ]);
    }
    /**
     * Sleep for specified milliseconds.
     */
    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
    /**
     * Context manager pattern for automatic cleanup.
     *
     * @param agent Agent to execute
     * @param fn Function to execute with the task
     * @param config Task configuration
     * @returns Result from the function
     *
     * @example
     * ```typescript
     * const result = await Task.withTask(agent, async (task) => {
     *   return await task.execute(message);
     * });
     * ```
     */
    static async withTask(agent, fn, config) {
        const task = new Task(agent, config);
        try {
            return await fn(task);
        }
        finally {
            await task.cleanup();
        }
    }
}
exports.Task = Task;
/**
 * Timeout error for task execution.
 */
class TimeoutError extends Error {
    constructor(message) {
        super(message);
        this.name = 'TimeoutError';
    }
}
exports.TimeoutError = TimeoutError;
/**
 * Execute a task with automatic cleanup.
 *
 * Convenience function that wraps Task creation, execution, and cleanup.
 *
 * @param agent Agent to execute
 * @param message Input message
 * @param config Task configuration
 * @returns Agent response
 *
 * @example
 * ```typescript
 * const result = await executeTask(
 *   agent,
 *   createMessage('user', 'Summarize this document'),
 *   { timeout: 30000, retries: 2 }
 * );
 * ```
 */
async function executeTask(agent, message, config) {
    return Task.withTask(agent, async (task) => task.execute(message), config);
}
